man/mknet.8
.TH mknet 8 "Version 0.3: 16 Dec 2020" "MKVM suite manuals" "mkvm collection"
.SH NAME
mknet \- Make Virtual Network
.SH SYNOPSIS
.B mknet
[
.B action
] [
.B options
]

.B cbclocale
[
.B -alhv
] [
.B -b <domain> -g <network range> -n <network> -u <uri> -d
]

.SH DESCRIPTION
\fBmknet\fP creates a virtual network in a libvirt host. You can also add the
domain into the cmdb for \fBcbcdomain\fP and \fBdnsa\fP.

.SH OPTIONS
.B Actions
.IP "-a,  --add"
add a network
.IP "-l,  --list"
list the active networks
.IP "-h,  --help"
display a short help message
.IP "-v,  --version"
display the program version
.PP
.B Details
.IP "-b,  --domain, --build-domain \fBdomain\fP"
domain name to be associated with the network. Also, the name of the
\fBcbcdomain\fP and \fBdnsa\fP domain.
.IP "-g,  --range \fBrange/prefix\fP"
IP Address range and prefix (x.x.x.x/x)
.IP "-n,  --name \fBname\fP"
Name for the network. This will also be the name of the network interface on
the virtualisation host
.IP "-u,  --uri \fBuri\fB"
Libvirt connection URI
.IP "-d,  --cmdb"
Add into \fBcmdb\fP for \fBcbcdomain\fP and \fBdnsa\fP
.PP
.SH FILES
.I /etc/cmdb/cmdb.conf
.PP
.I /etc/mkvm/mkvm.conf
.RS
System wide configuration files for cmdb and mkvm suites
.PP
.RE
.I ~/.cmdb.conf
.PP
.I ~/.mkvm.conf
.RS
User configuration files for cmdb and mkvm suites. See
.BR cmdb.conf (5)
.BR mkvm.conf (5)
for further details
.RE
.SH ENVIRONMENT
HOME is used to locate the user configuration files.
.SH AUTHOR
Iain M. Conochie <iain-at-thargoid-dot-co-dot-uk>
.SH "SEE ALSO"
.BR dnsa(8)
.BR cbcdomain(8)
.BR mkvm(8)
.BR mksp(8)


NEWS
---- Thursday 04th December 2014 --

Added the --enable-cbc check to configure

---- Wedensday 03rd December 2014 -

Updated dnsa so we can remove glue zones. Also added ability to add a .
at the end of the NS name if it contains the FQDN - otherwise this will
append the domain name to the end.

---- Sunday 30th November 2014 ----

Updated the autoconf configure.ac file - using the m4 directory to store
macros to check for libraries installed. This will be of use when we
start to add more libraries to be linked.

---- Monday 11th February 2012 ----

Release version 0.2.1 contains the new look cmdb. The database interface
has been modularised so that either sqlite or mysql can be used, and
also starting to use prepared statements.
There have been some other code clean ups too.
The 0.2.x branch will concentrate on the modularised structure for 
cbc and dnsa.

---- Friday 25th January 2013 -----

To celebrate Burns night this is the first release of the cmdb programs
using autoconf


src/mknet.c
/*
 *
 *  mknet : make network
 *  Copyright (C) 2018 - 2020  Iain M Conochie <iain-AT-thargoid.co.uk>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 *  mknet.c
 *
 *  main and command line parsing functions for mknet
 *
 */

#include <config.h>
#include <configmake.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>
#ifdef HAVE_GETOPT_H
# include <getopt.h>
#endif // HAVE_GETOPT_H
#ifdef HAVE_REGEX_H
# include <regex.h>
#endif // HAVE_REGEX_H
#ifdef HAVE_WORDEXP_H
# include <wordexp.h>
#endif // HAVE_WORDEXP_H
#include <ailsacmdb.h>
#include <ailsasql.h>
#include "virtual.h"

const char *prog = "mknet";

static int
parse_mknet_command_line(int argc, char *argv[], ailsa_mkvm_s *vm);

static void
display_usage(void);

int
main(int argc, char *argv[])
{
        int retval = 0;
        ailsa_mkvm_s *vm = ailsa_calloc(sizeof(ailsa_mkvm_s), "vm in main");
	ailsa_cmdb_s *cmdb = ailsa_calloc(sizeof(ailsa_cmdb_s), "cmdb in main");

	parse_mkvm_config(vm);
	parse_cmdb_config(cmdb);
	if ((retval = parse_mknet_command_line(argc, argv, vm)) != 0) {
                display_usage();
		goto cleanup;
        }
        switch(vm->action) {
        case CMDB_LIST:
                retval = ailsa_list_networks(vm);
                break;
        case AILSA_DISPLAY_USAGE:
                display_usage();
                break;
        case CMDB_ADD:
                retval = ailsa_add_network(cmdb, vm);
                break;
        case AILSA_VERSION:
                ailsa_syslog(LOG_INFO, "%s: version %s", prog, VERSION);
                break;
        }

        cleanup:
                ailsa_clean_mkvm(vm);
		ailsa_clean_cmdb(cmdb);
                return retval;
}

static int
parse_mknet_command_line(int argc, char *argv[], ailsa_mkvm_s *vm)
{
        int retval, opt;
        const char *optstr = "adhlvb:g:n:u:";

        retval = 0;
#ifdef HAVE_GOTOPT_H
        int index;
        struct option opts[] = {
                {"domain",              required_argument,      NULL,   'b'},
                {"build-domain",        required_argument,      NULL,   'b'},
                {"uri",                 required_argument,      NULL,   'u'},
                {"network",             required_argument,      NULL,   'n'},
                {"range",               required_argument,      NULL,   'g'},
                {"add",                 no_argument,            NULL,   'a'},
                {"cmdb",                no_argument,            NULL,   'd'},
                {"list",                no_argument,            NULL,   'l'},
                {"help",                no_argument,            NULL,   'h'},
                {"version",             no_argument,            NULL,   'v'},
                {NULL,                  0,                      NULL,   0}
        };
        while ((opt = getopt_long(argc, argv, optstr, opts, &index)) != -1)
#else
	while ((opt = getopt(argc, argv, optstr)) != -1)
#endif // HAVE_GETOPT_H
        {
                switch(opt) {
                case 'a':
                        vm->action = CMDB_ADD;
                        break;
                case 'l':
                        vm->action = CMDB_LIST;
                        break;
                case 'h':
                        vm->action = AILSA_DISPLAY_USAGE;
                        break;
                case 'v':
                        vm->action = AILSA_VERSION;
                        break;
                case 'g':
                        if ((retval = get_ip_addr_and_prefix(optarg, &(vm->range), &(vm->prefix))) != 0) {
                                ailsa_syslog(LOG_ERR, "Cannot split network range argument");
                                return retval;
                        }
                        break;
                case 'b':
                        if (strlen(optarg) >= DOMAIN_LEN)
                                ailsa_syslog(LOG_ERR, "Domain trimmed to 255 charaters");
                        if (vm->domain)
                                snprintf(vm->domain, DOMAIN_LEN, "%s", optarg);
                        else
                                vm->domain = strndup(optarg, DOMAIN_LEN);
                        break;
                case 'n':
                        if (strlen(optarg) >= CONFIG_LEN)
                                ailsa_syslog(LOG_ERR, "Network trimmed to 255 characters");
                        if (vm->network)
                                snprintf(vm->network, CONFIG_LEN, "%s", optarg);
                        else
                                vm->network = strndup(optarg, CONFIG_LEN);
                        break;
                case 'u':
                        if (strlen(optarg) >= CONFIG_LEN)
                                ailsa_syslog(LOG_ERR, "URI Trimmed to 255 characters");
                        if (vm->uri)
                                snprintf(vm->uri, CONFIG_LEN, "%s", optarg);
                        else
                                vm->uri = strndup(optarg, CONFIG_LEN);
                        break;
		case 'd':
			vm->cmdb = AILSA_CMDB_ADD;
			break;
                default:
			ailsa_syslog(LOG_ERR, "Unknown option %c\n", opt);
			return 1;
			break;
                }
        }
        if (vm->action == 0)
		retval = AILSA_NO_ACTION;
        if (!(vm->uri))
                retval = AILSA_NO_URI;
        return retval;
}

static void
display_usage(void)
{

        printf("%s: make virtual network %s\n", prog, VERSION);
        printf("Usage:\t");
        printf("%s <action> <options>\n", prog);
        printf("Actions\n");
        printf("\t-a: add\n");
        printf("\t-l: list\n");
        printf("\t-h: help\n");
        printf("\t-v: version\n");
        printf("Options\n");
        printf("\t-b <domain>: Build domain for the network\n");
        printf("\t-g <network/prefix>: Network range\n");
        printf("\t-u <uri>: Connection URI for libvirtd\n");
        printf("\t-n <network>: Name of the libvirt network\n");
        printf("\t-d Add network to cmdb\n");
}
